include/stl2/detail/iterator/any_iterator.hpp

```diff
@@ -0,0 +1,230 @@
+// cmcstl2 - A concept-enabled C++ standard library
+//
+//  Copyright Eric Niebler 2016
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/caseycarter/cmcstl2
+//
+#ifndef STL2_DETAIL_ITERATOR_ANY_ITERATOR_HPP
+#define STL2_DETAIL_ITERATOR_ANY_ITERATOR_HPP
+
+#include <new>
+#include <atomic>
+#include <stl2/type_traits.hpp>
+#include <stl2/detail/fwd.hpp>
+#include <stl2/detail/swap.hpp>
+#include <stl2/detail/concepts/core.hpp>
+#include <stl2/detail/iterator/basic_iterator.hpp>
+#include <stl2/detail/iterator/concepts.hpp>
+
+STL2_OPEN_NAMESPACE {
+	namespace __any_iterator {
+        enum class op { copy, move, nuke, bump, comp, rval };
+
+        struct counted {
+            std::atomic<long> cnt{ 1 };
+        };
+        InputIterator{I} struct shared_iterator : counted {
+            shared_iterator(I i) : it(std::move(i)) {}
+            I it;
+        };
+
+        union blob {
+            counted* big;
+            std::aligned_storage_t<2 * sizeof(void *)> tiny;
```

Comment:
Iterators smaller than or equal to two pointers are stored in-situ. The size of `any_input_iterator` overall is 4 pointers. We can argue if these are the right thresholds.